ci/codecov.sh

```diff
@@ -23,51 +23,69 @@ set -ex
+            # export PATH=${LLVM_COV_PATH}:${PATH}
+            # GCOV=llvm-cov
+        elif [ "${COMPILER_FAMILY}" == "gcc" ]; then
+            # cutting "gcc" above yields "gcc" for both values if no version present
+            if [ "${COMPILER_EDITION}" == "${COMPILER_FAMILY}" ]; then
+                GCOV=gcov
+            else
+                GCOV=gcov-${COMPILER_EDITION}
             fi
+        else
+            echo "Cannot determine GCOV."
+            exit 1
         fi
-        GCOV=gcov-${ver}
     fi
 
     # install the latest lcov we know works
     rm -rf /tmp/lcov
     cd /tmp
-    git clone --depth 1 -b v1.14 https://github.com/linux-test-project/lcov.git
+    git clone --depth 1 -b v1.15 https://github.com/linux-test-project/lcov.git
     export PATH=/tmp/lcov/bin:$PATH
     command -v lcov
     lcov --version
 
     # switch back to the original source code directory
     cd $BOOST_CI_SRC_FOLDER
-    : "${LCOV_BRANCH_COVERAGE:=1}" # Set default
+    : "${LCOV_BRANCH_COVERAGE:=1}" # On by default, job can override but advise against it
 
     # coverage files are in ../../b2 from this location
-    lcov --gcov-tool=$GCOV --rc lcov_branch_coverage=${LCOV_BRANCH_COVERAGE} --base-directory "$BOOST_ROOT/libs/$SELF" --directory "$BOOST_ROOT" --capture --output-file all.info
-    # dump a summary on the console
-    lcov --list all.info
+    lcov --rc lcov_branch_coverage=${LCOV_BRANCH_COVERAGE} --gcov-tool=${GCOV} --directory "${BOOST_ROOT}" --capture --output-file all.info
+    lcov --rc lcov_branch_coverage=${LCOV_BRANCH_COVERAGE} --list all.info  # mostly for debug but interesting to see everything captured
 
-    # all.info contains all the coverage info for all projects - limit to ours
-    # first we extract the interesting headers for our project then we use that list to extract the right things
+    # this is tricky; not all of the headers in the project go into a directory with the project
+    # name, so we have to itemize all of them and extract them, along with anything in the project
+    # directory including tests - after this we drop coverage on test files to finalize things
     for f in `for f in include/boost/*; do echo $f; done | cut -f2- -d/`; do echo "*/$f*"; done > /tmp/interesting
     echo headers that matter:
     cat /tmp/interesting
-    xargs -L 999999 -a /tmp/interesting lcov --gcov-tool=$GCOV --rc lcov_branch_coverage=${LCOV_BRANCH_COVERAGE:-1} --extract all.info {} "*/libs/$SELF/*" --output-file coverage.info
+    xargs -L 999999 -a /tmp/interesting lcov --rc lcov_branch_coverage=${LCOV_BRANCH_COVERAGE} --extract all.info {} "*/libs/$SELF/*" --output-file "${SELF}-all.info"
+    lcov --rc lcov_branch_coverage=${LCOV_BRANCH_COVERAGE} --list "${SELF}-all.info"
+
+    # drop coverage on test files
+    lcov --rc lcov_branch_coverage=${LCOV_BRANCH_COVERAGE} --remove "${SELF}-all.info" "*/${SELF}/test/*" -o "coverage.info"
 
-    # dump a summary on the console - helps us identify problems in pathing
```

Comment:
I'd keep the extended comment. Provides reasoning WHY we dump the summary, otherwise the comment wouldn't be required and someone might remove the code.